-/Kconfig/1.1.1.1/Thu Dec  1 11:53:50 2005//
-/Makefile/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /config.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /dummy_hcd.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
-/epautoconf.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
-/ether.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /file_storage.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
-/gadget_chips.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /goku_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /goku_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /inode.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
@@ -23,4 +18,9 @@
 /serial.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /usbstring.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
 /zero.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/Kconfig/1.1.1.1/Tue Apr 25 06:14:08 2006//
+/Makefile/1.1.1.1/Tue Apr 25 06:14:08 2006//
+/epautoconf.c/1.1.1.1/Tue Apr 25 06:14:08 2006//
+/ether.c/1.1.1.1/Tue Apr 25 06:14:09 2006//
+/gadget_chips.h/1.1.1.1/Tue Apr 25 06:14:10 2006//
 D
diff -Naru gadget_old/CVS/Entries.orig gadget_new/CVS/Entries.orig
--- gadget_old/CVS/Entries.orig	1969-12-31 16:00:00.000000000 -0800
+++ gadget_new/CVS/Entries.orig	2006-04-26 11:37:48.155040000 -0700
@@ -0,0 +1,26 @@
+/config.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/dummy_hcd.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/file_storage.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/goku_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/goku_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/inode.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/lh7a40x_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/lh7a40x_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/ndis.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/net2280.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/net2280.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/omap_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/omap_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/pxa2xx_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/pxa2xx_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/rndis.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/rndis.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/serial.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/usbstring.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/zero.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
+/Kconfig/1.1.1.1/Tue Apr 25 06:14:08 2006//
+/Makefile/1.1.1.1/Tue Apr 25 06:14:08 2006//
+/epautoconf.c/1.1.1.1/Tue Apr 25 06:14:08 2006//
+/ether.c/1.1.1.1/Tue Apr 25 06:14:09 2006//
+/gadget_chips.h/1.1.1.1/Tue Apr 25 06:14:10 2006//
+D
diff -Naru gadget_old/CVS/Entries.rej gadget_new/CVS/Entries.rej
--- gadget_old/CVS/Entries.rej	1969-12-31 16:00:00.000000000 -0800
+++ gadget_new/CVS/Entries.rej	2006-04-26 11:37:48.194040000 -0700
@@ -0,0 +1,55 @@
+***************
+*** 1,26 ****
+- /Kconfig/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /Makefile/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /config.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /dummy_hcd.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /epautoconf.c/1.1.1.1/Mon Apr 24 23:05:25 2006//
+- /ether.c/1.1.1.1/Mon Apr 24 23:05:27 2006//
+- /file_storage.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /gadget_chips.h/1.1.1.1/Mon Apr 24 23:05:27 2006//
+- /goku_udc.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /goku_udc.h/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /inode.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /lh7a40x_udc.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /lh7a40x_udc.h/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /ndis.h/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /net2280.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /net2280.h/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /omap_udc.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /omap_udc.h/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /pxa2xx_udc.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /pxa2xx_udc.h/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /rndis.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /rndis.h/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /serial.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /usbstring.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+- /zero.c/1.1.1.1/Mon Apr 24 23:03:21 2006//
+  D
+--- 1,26 ----
++ /config.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /dummy_hcd.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /file_storage.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /goku_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /goku_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /inode.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /lh7a40x_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /lh7a40x_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /ndis.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /net2280.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /net2280.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /omap_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /omap_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /pxa2xx_udc.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /pxa2xx_udc.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /rndis.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /rndis.h/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /serial.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /usbstring.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /zero.c/1.1.1.1/Thu Dec  1 11:53:50 2005//
++ /Kconfig/1.1.1.1/Mon Apr 24 22:43:46 2006//
++ /Makefile/1.1.1.1/Mon Apr 24 22:43:46 2006//
++ /epautoconf.c/1.1.1.1/Mon Apr 24 22:43:46 2006//
++ /ether.c/1.1.1.1/Mon Apr 24 22:43:47 2006//
++ /gadget_chips.h/1.1.1.1/Mon Apr 24 22:43:47 2006//
+  D
diff -Naru gadget_old/vox160_udc.c gadget_new/vox160_udc.c
--- gadget_old/vox160_udc.c	2006-04-26 12:09:24.914242000 -0700
+++ gadget_new/vox160_udc.c	2006-04-26 11:37:48.307041000 -0700
@@ -1076,6 +1076,8 @@
 	struct vox160_udc *udc = s->private;
 	unsigned long flags;
 	int i = 0;
+	struct vox160_ep *ep;
+	struct vox160_request *req;
 
 	spin_lock_irqsave(&udc->lock, flags);
 
@@ -1210,7 +1212,21 @@
 		tmp = readw(&udc->usbd_epadr[i]->reg);
 		seq_printf(s, "usbd_epadr[%d]  0x%04x \n", i, (tmp & 0x0fff));
 		tmp = readw(&udc->usbd_eplen[i]->reg);
-		seq_printf(s, "usbd_eplen[%d]  0x%x \n\n", i, (tmp));
+		seq_printf(s, "usbd_eplen[%d]  0x%x \n", i, (tmp));
+
+		/* dump endpoint queues */
+		ep = &udc->ep[i];
+		if (list_empty(&ep->queue)) {
+			seq_printf(s, "%s nothing queued \n\n", ep->ep.name);
+		} else {
+			list_for_each_entry(req, &ep->queue, queue) {
+				seq_printf(s, "%s req %p len %d:%d buf %p \n\n",
+					   ep->ep.name, &req->req,
+					   req->req.length, req->req.actual,
+					   req->req.buf);
+			}
+		}
+
 	}
 	seq_printf(s, "Total irqs  %ld \n", udc->irqs);
 	if (udc->sof_irqs)
@@ -1723,6 +1739,27 @@
 	VDBG(dev, "(%s:%d)\n", __FUNCTION__, __LINE__);
 }
 
+static inline void
+queue_advance(struct vox160_ep *ep)
+{
+	struct vox160_request *req = NULL;
+	int total;
+
+	if (unlikely(list_empty(&ep->queue)))
+		return;
+
+	req = list_entry(ep->queue.next, struct vox160_request, queue);
+
+	if (req) {
+		total = min(ep->max_buffer_len,
+			    (u16) (req->req.length - req->req.actual));
+		if (ep->is_in)
+			start_in_xfer(ep, req->req.buf, total);
+		else
+			start_out_xfer(ep, req->req.buf, total);
+	}
+}
+
 /* send packet data from the endpoint's fifo */
 static int
 write_fifo(struct vox160_ep *ep, struct vox160_request *req)
@@ -1770,6 +1807,7 @@
 			(u16) (req->req.length - req->req.actual));
 		status =
 		    start_in_xfer(ep, req->req.buf + req->req.actual, total);
+
 		VDBG(dev, "(%s:%d):%s %p BC len %d act %d\n", __FUNCTION__,
 		     __LINE__, ep->ep.name, &req->req, req->req.length,
 		     req->req.actual);
@@ -1786,6 +1824,9 @@
 		     req->req.actual);
 
 		done(ep, req, 0);
+
+		if (!list_empty(&ep->queue))
+			queue_advance(ep);
 	}
 	return is_done;
 }
@@ -1873,7 +1914,6 @@
 	struct vox160_udc *dev = ep->dev;
 	u16 ep_int_status, eplen;
 	int status = 0;
-	struct vox160_request r;
 
 	buf = req->req.buf + req->req.actual;
 	bufferspace = req->req.length - req->req.actual;
@@ -1903,7 +1943,6 @@
 		     req->req.length, req->req.actual,
 		     ep->current_max_buffer_len, eplen);
 
-		r.req.length = ep->current_max_buffer_len;
 		status =
 		    start_out_xfer(ep, NULL,
 				   min(ep->current_max_buffer_len,
@@ -1931,12 +1970,11 @@
 		     req->req.length, req->req.actual,
 		     ep->current_max_buffer_len, eplen);
 
-		r.req.length = ep->max_buffer_len;
-		status = start_out_xfer(ep, NULL, ep->max_buffer_len);
 		done(ep, req, 0);
-	}
 
-	return is_done;
+		if (!list_empty(&ep->queue))
+			queue_advance(ep);
+	}
 }
 
 static int
